### ISSUE TITLE
Questions about checkpoints

### ISSUE THREAD

**Issue by hyn1426536274** [mli0603/openpi-comet#7]
Thank you for your excellent work.

I would like to confirm that your main innovation is the expansion of official training data, and you have not made any changes to the model architecture or inference process, have you. (I do not mean to belittle your work. Actually, it's quite interesting)

---
Besides, I have some questions that I really look forward to hearing from you.

1. Are the primary pre-trained checkpoints provided in the repository (pi05-b1kpt12-cs32/pi05-b1kpt50-cs32) fine-tuned with RFT? Is pi05-b1kpt50-cs32 the best-performing pre-trained checkpoint? Have you tested the performance(scores) of this checkpoint? If not, which unified pre-trained model currently yields the optimal performance and what are its corresponding scores? What are the differences between pi05-pt50-pretrain-50k and pi05-b1kpt50-cs32, and which one is superior?
2. What is the relationship between the step count (15k/50k) and the number of trajectories (1M+)? As indicated in the report, for the pt50 pre-trained checkpoints (e.g., pi05-b1kpt50-cs32), was it trained for a total of 50k steps across 50 tasks (i.e., 1k steps per task)? And if fine-tuning is performed for a specific task, is the training conducted for 15k to 20k steps on that individual task?
3. (Regarding task IDs in the model zoo table) If model A contains a task ID x in the model zoo table, does it indicate that model A performs the best on task x (among all models). Taking pi05-pt12-pretrain-75k as an example, is my understanding correct that this model was pre-trained for 75k steps across 12 tasks? If so, why does pi05-pt50-pretrain-40k correspond to 21 task IDs, whereas pi05-pt50-pretrain-50k only corresponds to 2?
4. Has the well-performing unified pre-trained model with a q_score of 0.31, which was mentioned in other issue, been released? If yes, which specific model is it?
5. Do the total training steps of the models denoted as #pt7, #pt10, and #pt50 in the report remain consistent? Have you attempted to directly train a single task (e.g., turning on a radio) using the original pi05 base model (without any pre-training)? If yes, what were the resulting performance outcomes?

I’m new to this field, so please excuse my naive questions. Looking forward to hearing back from you again.

**Comment by hyn1426536274** [comment]
I am fine-tuning the pre-trained pi0.5 model (`pi05-b1kpt50-cs32`) on the "turning_on_radio" task. Although the loss decreases during training, the evaluation results are unsatisfactory. The fine-tuned model performs significantly worse than the original pre-trained model: the robot exhibits severe high-frequency jittering and fails to execute any effective movements.

For the behavior1k demos dataset, I only downloaded the `turning_on_radio` subset. I have tried one task (turn on radio) and 50 tasks (all tasks) for norm_stats.json. But the training results seem to be similar.

Here is the loss curve:
<img width="780" height="383" alt="Image" src="https://github.com/user-attachments/assets/ff8dd5f4-5c0e-40cd-9f6d-0d756a482d2d" />

Here is the TrainConfig:

```python
TrainConfig(
name="pi05_b1k-turning_on_radio", 
exp_name="openpi",
project_name="B1K",
model=pi0_config.Pi0Config(pi05=True, action_horizon=32),
data=LeRobotB1KDataConfig(
repo_id="behavior-1k/2025-challenge-demos",
base_config=DataConfig(
prompt_from_task=True,
episodes_index=list(range(200)),
behavior_dataset_root="path/to/dataset",
tasks=["turning_on_radio"],
fine_grained_level=0,  # 0: global instruction, 1: subtask instruction, 2: skill instruction
),
),
weight_loader=weight_loaders.CheckpointWeightLoader(
"/path/to/torch/pi05-b1kpt50-cs32"
),
num_train_steps=10_000,
lr_schedule=_optimizer.CosineDecaySchedule(
peak_lr=2.5e-5,
decay_steps=30_000,
),
freeze_filter=pi0_config.Pi0Config(pi05=True, action_horizon=32).get_freeze_filter(),
ema_decay=None,
checkpoint_base_dir="path/to/save/checkpoints",
assets_base_dir="path/to/assets/", 
num_workers=8,
batch_size=8,
),
``` 

Why does the loss decrease from such a high value (~1.5)? I thought this pre trained model should have already been trained on this task.
Why is the performance of the fine tuned model even much worse than that of the pre trained model in evaluation? (The robot keeps shaking and cannot make any effective movements)
By the way, how is your `norm_stats.json` calculated and why is the `norm_stats.json` I obtained using episodes from 50 tasks (10000 entries) different from the`norm_stats.json` in the assets folder of your pre trained model(`pi05-b1kpt50-cs32`).
Do you know what problem could be causing this? I am deeply sorry for causing you a lot of trouble.

**Comment by DelinQu** [comment]
> Thank you for your excellent work.
> 
> I would like to confirm that your main innovation is the expansion of official training data, and you have not made any changes to the model architecture or inference process, have you. (I do not mean to belittle your work. Actually, it's quite interesting)
> 
> Besides, I have some questions that I really look forward to hearing from you.
> 
> 1. Are the primary pre-trained checkpoints provided in the repository (pi05-b1kpt12-cs32/pi05-b1kpt50-cs32) fine-tuned with RFT? Is pi05-b1kpt50-cs32 the best-performing pre-trained checkpoint? Have you tested the performance(scores) of this checkpoint? If not, which unified pre-trained model currently yields the optimal performance and what are its corresponding scores? What are the differences between pi05-pt50-pretrain-50k and pi05-b1kpt50-cs32, and which one is superior?
> 2. What is the relationship between the step count (15k/50k) and the number of trajectories (1M+)? As indicated in the report, for the pt50 pre-trained checkpoints (e.g., pi05-b1kpt50-cs32), was it trained for a total of 50k steps across 50 tasks (i.e., 1k steps per task)? And if fine-tuning is performed for a specific task, is the training conducted for 15k to 20k steps on that individual task?
> 3. (Regarding task IDs in the model zoo table) If model A contains a task ID x in the model zoo table, does it indicate that model A performs the best on task x (among all models). Taking pi05-pt12-pretrain-75k as an example, is my understanding correct that this model was pre-trained for 75k steps across 12 tasks? If so, why does pi05-pt50-pretrain-40k correspond to 21 task IDs, whereas pi05-pt50-pretrain-50k only corresponds to 2?
> 4. Has the well-performing unified pre-trained model with a q_score of 0.31, which was mentioned in other issue, been released? If yes, which specific model is it?
> 5. Do the total training steps of the models denoted as #pt7, #pt10, and #pt50 in the report remain consistent? Have you attempted to directly train a single task (e.g., turning on a radio) using the original pi05 base model (without any pre-training)? If yes, what were the resulting performance outcomes?
> 
> I’m new to this field, so please excuse my naive questions. Looking forward to hearing back from you again.

Hi! Sorry for the delayed reply — thank you very much for your patience.

🏆 Post-Challenge update (TL;DR):
After the competition, we refined our training and post-training pipeline and achieved a Q-score of 0.345 on the Public Validation set, while compressing the released model zoo down to only two unified pretrained checkpoints:
[Model Zoo (two pretrained models)](https://github.com/mli0603/openpi-comet?tab=readme-ov-file#model-zoo)

So most of the older checkpoint naming / long model lists are no longer relevant.
|   Model Name | Discription                          | HuggingFace URL                                                                                                                                                                                                                                          |
|----------:|:----------------------------------------|:---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| pi05-b1kpt12-cs32 | Pretrained Model in tasks `0,1,6,17,18,22,30,32,34,35,40,45` with action chunk size 32                         | [openpi_comet/pi05-b1kpt12-cs32](https://huggingface.co/sunshk/openpi_comet/tree/main/pi05-b1kpt12-cs32)                                                                  |
| pi05-b1kpt50-cs32 | Pretrained Model in tasks `0-49` with action chunk size 32                                                     | [openpi_comet/pi05-b1kpt50-cs32](https://huggingface.co/sunshk/openpi_comet/tree/main/pi05-b1kpt50-cs32)                                                                  |

A few clarifications:

1. Architecture / inference:
We did not change the core OpenPi model architecture or the inference/serving procedure. Our improvements mainly come from (i) multi-task pretraining data & recipe and (ii) post-training (RFT / Post-training++).

2. How the reported score is produced (0.345):
The reported Q-score/ACC is obtained by merging the evaluation results from these two pretrained models (we ran the evaluation once under a unified config).
These two models are pretrained from OpenPi 05 using:

https://huggingface.co/datasets/behavior-1k/2025-challenge-demos

https://huggingface.co/datasets/delinqu/comet-1.5k

3. Which checkpoint to use for single-task fine-tuning:
- If you are fine-tuning a single task, and the task is in the pt12 task list, start from pi05-b1kpt12-cs32 (often better specialization).

- Otherwise, pi05-b1kpt50-cs32 is strongly recommended for general coverage.

3. Why pt12 can outperform pt50 on some tasks:
It’s common that pt12 performs better than pt50 on the tasks it covers, because it is more specialized and may suffer less from capacity dilution. A plausible reason is that the Pi0.5 model capacity is limited, so spreading across 50 tasks can reduce per-task specialization.

4. About training steps vs. trajectories / scaling:
The total global steps are not fixed across settings; we roughly scale the training budget with the number of episodes.
As a rule of thumb, if a single-task run uses something like 8 GPUs, batch size 64, ~15k steps, then a 12-task pretraining run would use a larger budget roughly proportional to the number of tasks/data volume (not “1k steps per task”, but global steps on a mixed multi-task sampler).

5. Single-task SFT from the original Pi0.5 base (no B1K pretraining):
Yes — we tried directly fine-tuning the official Pi0.5 base on single tasks. The results were very poor: only task 0/1 reached <30% success, while most other tasks were near zero. This strongly highlights the importance of BEHAVIOR-1K-style pretraining.

Happy to provide more guidance if you tell me which specific task(s) you’re interested in — I can suggest the best starting checkpoint and a practical fine-tuning recipe.

**Comment by DelinQu** [comment]
> I am fine-tuning the pre-trained pi0.5 model (`pi05-b1kpt50-cs32`) on the "turning_on_radio" task. Although the loss decreases during training, the evaluation results are unsatisfactory. The fine-tuned model performs significantly worse than the original pre-trained model: the robot exhibits severe high-frequency jittering and fails to execute any effective movements.
> 
> For the behavior1k demos dataset, I only downloaded the `turning_on_radio` subset. I have tried one task (turn on radio) and 50 tasks (all tasks) for norm_stats.json. But the training results seem to be similar.
> 
> Here is the loss curve: <img alt="Image" width="780" height="383" src="https://private-user-images.githubusercontent.com/121924201/528469234-ff8dd5f4-5c0e-40cd-9f6d-0d756a482d2d.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Njc1MDQ0MzUsIm5iZiI6MTc2NzUwNDEzNSwicGF0aCI6Ii8xMjE5MjQyMDEvNTI4NDY5MjM0LWZmOGRkNWY0LTVjMGUtNDBjZC05ZjZkLTBkNzU2YTQ4MmQyZC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjYwMTA0JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI2MDEwNFQwNTIyMTVaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT00NjZkNjI4NzBkYjgyM2E1YWVlNmNlNGM3NmM2NjJkOGJlYTYxMWIxZDQ4ZWMwZWIyMmFlMWNjYzBmODc0NGU4JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.5bEpenGxhi-HtdE4Ec5mGz-UtGcdEogLaGUwqoTLlWc">
> 
> Here is the TrainConfig:
> 
> TrainConfig(
> name="pi05_b1k-turning_on_radio", 
> exp_name="openpi",
> project_name="B1K",
> model=pi0_config.Pi0Config(pi05=True, action_horizon=32),
> data=LeRobotB1KDataConfig(
> repo_id="behavior-1k/2025-challenge-demos",
> base_config=DataConfig(
> prompt_from_task=True,
> episodes_index=list(range(200)),
> behavior_dataset_root="path/to/dataset",
> tasks=["turning_on_radio"],
> fine_grained_level=0,  # 0: global instruction, 1: subtask instruction, 2: skill instruction
> ),
> ),
> weight_loader=weight_loaders.CheckpointWeightLoader(
> "/path/to/torch/pi05-b1kpt50-cs32"
> ),
> num_train_steps=10_000,
> lr_schedule=_optimizer.CosineDecaySchedule(
> peak_lr=2.5e-5,
> decay_steps=30_000,
> ),
> freeze_filter=pi0_config.Pi0Config(pi05=True, action_horizon=32).get_freeze_filter(),
> ema_decay=None,
> checkpoint_base_dir="path/to/save/checkpoints",
> assets_base_dir="path/to/assets/", 
> num_workers=8,
> batch_size=8,
> ),
> Why does the loss decrease from such a high value (~1.5)? I thought this pre trained model should have already been trained on this task. Why is the performance of the fine tuned model even much worse than that of the pre trained model in evaluation? (The robot keeps shaking and cannot make any effective movements) By the way, how is your `norm_stats.json` calculated and why is the `norm_stats.json` I obtained using episodes from 50 tasks (10000 entries) different from the`norm_stats.json` in the assets folder of your pre trained model(`pi05-b1kpt50-cs32`). Do you know what problem could be causing this? I am deeply sorry for causing you a lot of trouble.

Hi, thanks for the detailed report — and no worries at all.

The core issue is normalization.

1. Why does the loss start very high (~1.5)?
When fine-tuning from pi05-b1kpt50-cs32, you must inherit the normalization stats from the pretrained checkpoint.
If you recompute norm_stats.json, the original flow-matching trajectory is broken, which leads to high initial loss and unstable policies (e.g., jittering), even though the loss may decrease.

👉 Please do not pass any new norm_stats.json when resuming from a pretrained model.

2. Why can fine-tuning be worse than pretraining?
Single-task SFT does not always improve performance and can sometimes hurt generalization, especially for BEHAVIOR-1K tasks.
However, in your case the severe shaking is almost certainly caused by Issue (1). Please fix that first.

3. Why is your norm_stats.json different from ours?
Our released normalization stats are computed from official demos + RFT data, so differences are expected. This is another reason not to recompute stats during fine-tuning.

Once you reuse the pretrained normalization, the behavior should be much more stable. If not, feel free to share your config.

**Comment by hyn1426536274** [comment]
I have a general understanding, thank you very much for your patient clarification.
